ext/cache/01_cache.js

```diff
@@ -0,0 +1,261 @@
+
+  class CacheStorage {
+    constructor() {
+      return this;
+    }
+
+    async open(cacheName) {
+      const cacheId = await core.opAsync("op_cache_storage_open", cacheName);
+      return PromiseResolve(new Cache(cacheId));
+    }
+
+    async has(cacheName) {
+      return await core.opAsync("op_cache_storage_has", cacheName);
+    }
+
+    async delete(cacheName) {
+      return await core.opAsync("op_cache_storage_delete", cacheName);
+    }
+  }
+
+  class Cache {
+    #id;
+
+    constructor(cacheId) {
+      this.#id = cacheId;
+      return this;
+    }
+
+    /** See https://w3c.github.io/ServiceWorker/#dom-cache-put */
+    async put(request, response) {
+      // Step 1.
+      let innerRequest = null;
+      // Step 2.
+      if (request instanceof Request) {
+        innerRequest = toInnerRequest(request);
+      } else {
+        // Step 3.
+        try {
+          innerRequest = toInnerRequest(new Request(request));
+        } catch (error) {
+          return PromiseReject(error);
+        }
+      }
+      // Step 4.
+      const reqUrl = innerRequest.url();
+      if (!reqUrl.startsWith("http:") && !reqUrl.startsWith("https:")) {
+        const url = new URL(reqUrl);
+        if (url.protocol !== "http:" && url.protocol !== "https:") {
+          return PromiseReject(
+            new TypeError("Request url protocol must be http or https"),
+          );
+        }
+      }
+      if (innerRequest.method !== "GET") {
+        return PromiseReject(new TypeError("Request method must be GET"));
+      }
+      // Step 5.
+      const innerResponse = toInnerResponse(response);
+      // Step 6.
+      if (innerResponse.status === 206) {
+        return PromiseReject(
+          new TypeError("Response status must not be 206"),
+        );
+      }
+      // Step 7.
+      const varyHeader = getHeader(innerResponse.headerList, "vary");
+      if (varyHeader) {
+        const fieldValues = varyHeader.split(",").map((field) => field.trim());
+        for (const fieldValue of fieldValues) {
+          if (
+            fieldValue === "*"
+          ) {
+            return PromiseReject(
+              new TypeError("Vary header must not contain '*'"),
+            );
+          }
+        }
+      }
+
+      // Step 8.
+      if (innerResponse.body.unusable()) {
+        return PromiseReject(
+          new TypeError("Response body must not already used"),
+        );
+      }
+
+      // Step 9-11.
+      const rid = await core.opAsync(
+        "op_cache_put",
+        {
+          cacheId: this.#id,
+          requestUrl: innerRequest.url(),
+          responseHeaders: innerResponse.headerList,
+          requestHeaders: innerRequest.headerList,
+          responseHasBody: innerResponse.body !== null,
+          responseStatus: innerResponse.status,
+          responseStatusText: innerResponse.statusMessage,
+        },
+      );
+      if (innerResponse.body) {
+        const reader = innerResponse.body.stream.getReader();
```

Comment:
Not familiar with this section and haven't looked around too well at the moment (sorry), does the reader evoke user code?

If it's a possibility that this is a user-provided `ReadableStream`, then I'll double-check and cross-reference this spot with the spec.